nx/lib/nx.ex

```diff
@@ -3985,6 +4001,94 @@ defmodule Nx do
 
+  @doc """
+  Reshapes input tensors so that they are all vectorized with the same vectors.
+
+  For vectors with the same name to be compatible, they need to either
+  have the same size or one must be of size 1.
+
+  ## Examples
+
+  Two vectors of the same name are compatible if they have the same sizes or if either has size 1.
+
+      iex> x = Nx.tensor([1, 2, 3]) |> Nx.vectorize(:x)
+      iex> xy = Nx.tensor([[[5]], [[6]]]) |> Nx.vectorize(:y) |> Nx.vectorize(:x)
+      iex> [x, xy] = Nx.reshape_vectors([x, xy])
+      iex> x.vectorized_axes
+      [x: 3, y: 1]
+      iex> xy.vectorized_axes
+      [x: 1, y: 2]
+
+  The resulting tensors will all present the combined vectors in the
+  same order. In the example below, those are `:x`, `:y` and `:z`.
+  Note that the sizes will always be either `1` or the original size
+  of the vector for that tensor.
+
+      iex> x = Nx.tensor([1, 2, 3]) |> Nx.vectorize(:x)
+      iex> y = Nx.tensor([4]) |> Nx.vectorize(:y)
+      iex> xy = Nx.tensor([[[5]], [[6]]]) |> Nx.vectorize(:y) |> Nx.vectorize(:x)
+      iex> z = Nx.tensor([8]) |> Nx.vectorize(:z)
+      iex> [x, y, xy, z] = Nx.reshape_vectors([x, y, xy, z])
+      iex> x.vectorized_axes
+      [x: 3, y: 1, z: 1]
+      iex> y.vectorized_axes
+      [x: 1, y: 1, z: 1]
+      iex> xy.vectorized_axes
+      [x: 1, y: 2, z: 1]
+      iex> z.vectorized_axes
+      [x: 1, y: 1, z: 1]
+  """
+  @doc type: :shape
+  def reshape_vectors(tensors)
+
+  def reshape_vectors([tensor]), do: [tensor]
+
+  def reshape_vectors([first | tensors]) do
+    # For all tensors to be compatible, each pair also needs to be compatible
+    # This means that we can do a first pass accumulating axes into
+    # the first tensor, and then a second pass getting them all into their final shapes.
+    left =
+      Enum.reduce(tensors, first, fn right, left ->
+        {left, _, vectorized_axes} = devectorize_with_axes(left, to_tensor(right))
+        revectorize_and_validate_sizes(left, vectorized_axes, true)
+      end)
+
+    left = reshape(left, shape(first))
+
+    tail =
+      Enum.map(tensors, fn %{shape: inner_shape} = right ->
+        {_, right, vectorized_axes} = devectorize_with_axes(left, right)
+
+        same_order =
+          Enum.reduce_while(left.vectorized_axes, vectorized_axes, fn
+            {left_name, _}, [{right_name, _} | axes] ->
+              cond do
+                left_name != right_name ->
+                  {:halt, false}
+
+                axes == [] ->
+                  {:cont, true}
+
+                true ->
+                  {:cont, axes}
+              end
+          end)
+
+        vectorized =
+          if same_order do
+            revectorize_and_validate_sizes(right, vectorized_axes, true)
+          else
+            reorder_vectorized_axes(right, vectorized_axes, left.vectorized_axes)
+          end
+
+        # we also want to ensure the inner shape is the same
+
+        reshape(vectorized, inner_shape)
```

Comment:
Doesn't devectorize_with_axes already reshape?

Reply:
The problem is that it reshapes in a broadcast able way, so the inner shape for a scalar might become {1} for instance.

At least this is what happened during my tests.
Might be a bug elsewhere though